Added a new section for Deployment and testing guidelines for Developers (#365)
opendatahub-io/data-science-pipelines-operator: README.md

   2. [Using the API](#using-the-api)
5. [Cleanup](#cleanup)
   1. [Cleanup ODH Installation](#cleanup-odh-installation)
   2. [Cleanup Standalone Installation](#cleanup-standalone-installation)
6. [Run tests](#run-tests)
7. [Metrics](#metrics)
8. [Deployment and Testing Guidelines for Developers](#deployment-and-testing-guidelines-for-developers)

# Overview

Data Science Pipelines (DSP) allows data scientists to track progress as they
iterate over development of ML models. With DSP, a data scientist can create
workflows for data preparation, model training, model validation, and more.
They can create and track experiements to arrive at the best version of of
training data, model hyperparameters, model code, etc., and repeatably
rerun these experiments.

Data Science Pipelines is based on the upstream [Kubeflow Pipelines (KFP)][kfp]
project. We leverage the [kfp-tekton][kfp] project to run pipelines backed
by the Tekton (rather than Argo, which is the default choice in KFP). We
currently distributed version 1.x of KFP, and are working to support v2.

Data Scientists can use tools like the
[kfp-tekton SDK](https://github.com/kubeflow/kfp-tekton/blob/master/sdk/README.md)

> or [Elyra](https://github.com/elyra-ai/elyra) to author their workflows, and

Reviewer: this commit should not show up, you'll need to rebase over this change

[flipcoin code example]: https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/docs/example_pipelines/condition.py
[installodh]: https://opendatahub.io/docs/quick-installation
[kfp-tekton]: https://github.com/kubeflow/kfp-tekton

# Deployment and Testing Guidelines for Developers

To run the DSPO locally :

Clone the directory and execute the following command:

```bash

> go run main.go --config=$config.yaml

Reviewer 2: Could you list out contents for the `config.yaml` file?

```
To run the tests:
 
Execute `make test` or `make unittest` or `make functest` based on the level of testing that needs to be done.

To deploy DSPO as a developer :

> Follow the instructions from [here](#deploy-the-operator-standalone) to deploy the operator standalone.

Reviewer 2: You could also mention deploying the operator via ODH/kfdef

[flipcoin code example]: https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/docs/example_pipelines/condition.py
[installodh]: https://opendatahub.io/docs/quick-installation
[kfp-tekton]: https://github.com/kubeflow/kfp-tekton

> # Deployment and Testing Guidelines for Developers

Author: Thank you @HumairAK , will cover the suggested points.

> # Deployment and Testing Guidelines for Developers

Reviewer 3: We also should add subsections for:
- handling/running pre-commit
- building/generating manifests

> To run the DSPO locally :

Reviewer 3: We should also mention that you need to be logged-into the OpenShift cluster via `oc login`, and CRDs need to be installed (how do we do that?)

Clone the directory and execute the following command:

```bash
go run main.go --config=$config.yaml
```
Below is the sample config file(the tags for the images can be edited as required):
```bash
Images:

>   ApiServer: quay.io/opendatahub/ds-pipelines-api-server:main-0e8a011

Reviewer 3: we should update these images to use latest tags (rather than explicit SHAs/dev tags) where applicable

  Artifact: quay.io/opendatahub/ds-pipelines-artifact-manager:main-0e8a011
  OAuthProxy: registry.redhat.io/openshift4/ose-oauth-proxy:v4.12.0
  PersistentAgent: quay.io/opendatahub/ds-pipelines-persistenceagent:main-0e8a011
  ScheduledWorkflow: quay.io/opendatahub/ds-pipelines-scheduledworkflow:main-0e8a011
  Cache: registry.access.redhat.com/ubi8/ubi-minimal
  MoveResultsImage: registry.access.redhat.com/ubi8/ubi-micro
  MariaDB: registry.redhat.io/rhel8/mariadb-103:1-188
  MlmdEnvoy: quay.io/opendatahub/ds-pipelines-metadata-envoy:1.7.0
  MlmdGRPC: quay.io/opendatahub/ds-pipelines-metadata-grpc:1.0.0
  MlmdWriter: quay.io/opendatahub/ds-pipelines-metadata-writer:1.1.0
  CRDViewer: gcr.io/ml-pipeline/viewer-crd-controller:2.0.0-rc.2
  VisualizationServer: gcr.io/ml-pipeline/visualization-server:2.0.0-rc.2
```
To run the tests:
 

> Execute `make test` or `make unittest` or `make functest` based on the level of testing that needs to be done.

Reviewer 3: Let's expand on what each of these do.  example: functests are functional tests (what does that mean?), `make test` just runs both unittests and functests, etc

Below is the sample config file(the tags for the images can be edited as required):
```bash
Images:
  ApiServer: quay.io/opendatahub/ds-pipelines-api-server:latest
  Artifact: quay.io/opendatahub/ds-pipelines-artifact-manager:latest
  OAuthProxy: registry.redhat.io/openshift4/ose-oauth-proxy:v4.12.0
  PersistentAgent: quay.io/opendatahub/ds-pipelines-persistenceagent:latest
  ScheduledWorkflow: quay.io/opendatahub/ds-pipelines-scheduledworkflow:latest
  Cache: registry.access.redhat.com/ubi8/ubi-minimal
  MoveResultsImage: registry.access.redhat.com/ubi8/ubi-micro
  MariaDB: registry.redhat.io/rhel8/mariadb-103:1-188
  MlmdEnvoy: quay.io/opendatahub/ds-pipelines-metadata-envoy:latest
  MlmdGRPC: quay.io/opendatahub/ds-pipelines-metadata-grpc:latest
  MlmdWriter: quay.io/opendatahub/ds-pipelines-metadata-writer:latest

>   CRDViewer: gcr.io/ml-pipeline/viewer-crd-controller:2.0.0-rc.2

Reviewer 3: let's remove these 2 images, as I don't believe we'll be using introducing them now (from unrelated spikes/decisions during this sprint)

```

How to regenerate crd on api changes:

After making your API changes, run the following command to regenerate code and CRDs based on your updated API definitions:

```bash
make generate
```
Refer to kubebuilder docs [here][kubebuilder-docs] for more info.

How to run pre-commit tests:

Install pre-commit following the instructions [here][pre-commit-installation]. Before creating a PR, developers should run the following command which will auto fix any simple errors:

```bash
pre-commit run --all-files
```

How to do disable health checks when dev testing:

To disable the health checks set the values to true in the DSPA yaml file you apply. Refer to his sample file [here][dspa-yaml].

In certain scenarios, it may be necessary to disable health checks within our environment. When the DSPO is executed either locally or on a different cluster, the health checks can't reach the database and Object Store endpoints. Consequently, they remain unsuccessful, preventing the deployment of essential pipeline infrastructure components by the DSPA. To address this challenge, we have introduced the `disableHealthCheck` mechanism as a viable solution.

How to enable kfp ui and minio:

> Refer to this [sample][sample-yaml] yaml file to know how to enable kfp ui and minio. 

Reviewer 3: ```suggestion
Refer to this [sample][sample-yaml] yaml file for enabling the kfp ui and minio. 
```
